nautobot/extras/models/datasources.py

```diff
@@ -167,3 +174,79 @@ def sync(self, user, dry_run=False):
         if dry_run:
             return enqueue_git_repository_diff_origin_and_local(self, user)
         return enqueue_pull_git_repository_and_refresh_data(self, user)
+
+    @contextmanager
+    def clone_to_directory_context(self, path=None, branch=None, head=None, depth=0):
+        """
+        Context manager to perform a (shallow or full) clone of the Git repository in a temporary directory.
+
+        Args:
+            path (str, optional): The absolute directory path to clone into. If not specified, `tempfile.gettempdir()` will be used.
+            branch (str, optional): The branch to checkout. If not set, the GitRepository.branch will be used.
+            head (str, optional): Git commit hash to check out instead of pulling branch latest.
+            depth (int, optional): The depth of the clone. If set to 0, a full clone will be performed.
+
+        Returns:
+            Returns the absolute path of the cloned repo if clone was successful, otherwise returns None.
+        """
+
+        if branch and head:
+            raise ValueError("Cannot specify both branch and head")
+
+        path_name = None
+        try:
+            path_name = self.clone_to_directory(path=path, branch=branch, head=head, depth=depth)
+            yield path_name
+        finally:
+            # Cleanup the temporary directory
+            if path_name:
+                self.cleanup_cloned_directory(path_name)
+
+    def clone_to_directory(self, path=None, branch=None, head=None, depth=0):
+        """
+        Perform a (shallow or full) clone of the Git repository in a temporary directory.
+
+        Args:
+            path (str, optional): The absolute directory path to clone into. If not specified, `tempfile.gettempdir()` will be used.
+            branch (str, optional): The branch to checkout. If not set, the GitRepository.branch will be used.
+            head (str, optional): Git commit hash to check out instead of pulling branch latest.
+            depth (int, optional): The depth of the clone. If set to 0, a full clone will be performed.
+
+        Returns:
+            Returns the absolute path of the cloned repo if clone was successful, otherwise returns None.
+        """
+        if branch and head:
+            raise ValueError("Cannot specify both branch and head")
+
+        try:
+            path_name = tempfile.mkdtemp(dir=path, prefix=self.slug)
+        except PermissionError as e:
+            logger.error(f"Failed to create temporary directory at {path}: {e}")
+            raise e
+
+        if not branch:
+            branch = self.branch
+
+        try:
+            repo_helper = GitRepo(path_name, self.remote_url, depth=depth, branch=branch)
+            head, _ = repo_helper.checkout(branch, head)
```

Comment:
Shouldn't be needed if we're already cloning the specified branch?

Reply:
Is this what you mean?
```
repo_helper = GitRepo(path_name, self.remote_url, depth=depth, branch=branch)
if head:
     head, _ = repo_helper.checkout(branch, head)
```

Reply:
think so, yeah. Thanks.